Fast gemm for einsum

Abdurrahheem opened the pull request:
## This PR adds performance tests for Einsum Layer with FastGemm. See below results of performance test on different inputs

**Notation:**
- WX: windows10_x64
- MX: macos_x64
- MA: macos_arm64
- UX: ubuntu_x64
- UA: ubuntu_arm64

All data in ms (milliseconds).
Gemm is backend for matrix multiplication

---

Benchmarks: (arrow indicates increase in inference speed compared to [einsum with gemm](https://github.com/opencv/opencv/pull/24445))

| Equation                | Inputs Mat Dims                    | UX (ms)  | UA (ms) | MX (ms) | MA (ms) | WX (ms) |
|-------------------------|------------------------------------|----------|---------|---------|---------|---------|
| "ij, jk -> ik"          | [2, 3], [3,2]                      | 0.04 ± 0.00 | -       | -       | -       | -       |
| "ij, jk -> ik"          | [20, 30], [30,20]                  | 0.07 ± 0.00 | -       | -       | -       | -       |
| "ij, jk -> ik"          | [113, 127], [127,113]              | 1.17 ± 0.02 ↓ ~ 48%| -       | -       | -       | -       |
| "imkj, injs -> imnks"   | [1, 4, 7, 9], [1, 5, 9, 8]         | 0.10 ± 0.00  | -       | -       | -       | -       |
| "imkj, injs -> imnks"   | [1, 4, 70, 90], [1, 5, 90, 80]     | 5.75 ± 0.10 ↓ ~ 37%| -       | -       | -       | -       |
| "imkj, injs -> imnks"   | [1, 4, 73, 91], [1, 5, 91, 57]     | 5.58 ± 0.12 ↓ ~ 48%| -       | -       | -       | -       |
| "ij -> i"               | [30, 40]                           | 0.03 ± 0.00  | -       | -       | -       | -       |
| "ij -> i"               | [113, 374]                         | 0.13 ± 0.00  | -       | -       | -       | -       |
| "...ij -> ...i"         | [30, 40]                           | 0.03 ± 0.00  | -       | -       | -       | -       |
| "...ij -> ...i"         | [113, 374]                         | 0.13 ± 0.00  | -       | -       | -       | -       |
| "...ij, ...jk -> ...ik" | [40, 50], [50,80]                  | 0.26 ± 0.00 | -       | -       | -       | -       |
| "...ij, ...jk -> ...ik" | [47, 51], [51, 83]                 | 0.28 ± 0.01  | -       | -       | -       | -       |


### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

asmorkalov commented on modules/dnn/src/layers/einsum_layer.cpp:
```diff
@@ -38,6 +40,10 @@ static Mat batchwiseMatMul(
     const MatShape& input2ShapeOverride)
 {
+    // hangle this properly in the einsum layer
+    FastGemmOpt opt;
+    opt.init();
```
I propose to make it layer-wide, but not initialize each time.

fengyuentau commented:
How do I understand your collected results? Is it faster or slower compared with `cv::gemm`?

Abdurrahheem commented:
> How do I understand your collected results? Is it faster or slower compared with `cv::gemm`?

Sorry. I added description. Inference is faster with FastGemm

dkurt commented on modules/dnn/perf/perf_einsum.cpp:
```diff
@@ -0,0 +1,123 @@
+};
+
+static inline void PrintTo(const EinsumParams& params, ::std::ostream* os) {
+     (*os) << "Eqiation=" << params.equation << ", "
```
Equation

dkurt commented on modules/dnn/perf/perf_einsum.cpp:
```diff
@@ -0,0 +1,123 @@
+static inline void PrintTo(const EinsumParams& params, ::std::ostream* os) {
+     (*os) << "Eqiation=" << params.equation << ", "
+        << "InputSize=" << params.inputSize << ", "
+        << "OutputSize=" << params.outputSize << ", ";
```
I think that number of inputs and output is not very informative. Shapes are enough

dkurt commented on modules/dnn/perf/perf_einsum.cpp:
```diff
@@ -0,0 +1,123 @@
+// test cases
+static const EinsumParams testEinsumConfigs[] = {
+    // TODO: Add tests with one input after ellips merge
+    {"ij, jk -> ik", 2, 1,  {{2, 3}, {3, 2}}},
```
I think that test case is too fast to get reliable perf data

Abdurrahheem replied:
When these test were larges @fengyuentau suggested removing large ones. Which one should I follow?

dkurt replied:
It might depends. Please provide a raw output from performance test to see deviation.

dkurt commented on modules/dnn/perf/perf_einsum.cpp:
```diff
@@ -0,0 +1,123 @@
+    {"ij, jk -> ik", 2, 1,  {{20, 30}, {30, 20}}},
+    {"ij, jk -> ik", 2, 1,  {{113, 127}, {127, 113}}},
+
+    {"imkj, injs -> imnks", 2, 1,  {{1, 4, 7, 9}, {1, 5, 9, 8}}},
```
Same as above

dkurt commented on modules/dnn/perf/perf_einsum.cpp:
```diff
@@ -0,0 +1,123 @@
+
+    {"...ij, ...jk -> ...ik", 2, 1, {{40, 50}, {50, 80}}},
+    {"...ij, ...jk -> ...ik", 2, 1, {{47, 51}, {51, 83}}},
+};
```
inputSize seems redundant as there are input shapes

dkurt commented on modules/dnn/perf/perf_einsum.cpp:
```diff
@@ -0,0 +1,123 @@
+
+        input_names.emplace_back("input1");
+        input_names.emplace_back("input2");
+    }
```
Can be simplified:
```cpp
int id = net.addLayer(lp.name, lp.type, lp);
for (int i = 0; i < params.inputSize; ++i) {
  inputs.emplace_back(Mat(params.einsumInpShapes[i].size(), params.einsumInpShapes[i].data(), CV_32FC1));
  net.connect(0, i, id, i);
}
```

dkurt commented on modules/dnn/perf/perf_einsum.cpp:
```diff
@@ -0,0 +1,123 @@
+    for (int i = 0; i < input_names.size(); i++){
+        net.setInput(inputs[i], input_names[i]);
+    }
+    Mat out = net.forward();
```
This warmup forward is different from `forward` in TEST_CYCLE

Abdurrahheem replied:
Are you suggesting to remove the warm up? Sorry, did not catch it

asmorkalov replied:
Warmup and test cycle should do the same job.

dkurt commented on modules/dnn/src/layers/einsum_layer.cpp:
@Abdurrahheem, perf file disappeared in this commit

dkurt replied:
Oh, I see, it was already merged in https://github.com/opencv/opencv/pull/24445

asmorkalov commented:
armv7 neon (jetson tk1):
```
Geometric mean (ms)

                                                       Name of Test                                                        4.x-baseline-1 4.x-fastgemm-1 4.x-fastgemm-1
                                                                                                                                                               vs      
                                                                                                                                                         4.x-baseline-1
                                                                                                                                                           (x-factor)  
einsum::Layer_Einsum::Eqiation=...ij -> ...i, InputSize=1, OutputSize=1, InputShape={{30, 40}}                                 0.027          0.027           1.03     
einsum::Layer_Einsum::Eqiation=...ij -> ...i, InputSize=1, OutputSize=1, InputShape={{113, 374}}                               0.120          0.120           1.00     
einsum::Layer_Einsum::Eqiation=...ij, ...jk -> ...ik, InputSize=2, OutputSize=1, InputShape={{40, 50}, {50, 80}}               0.459          0.269           1.71     
einsum::Layer_Einsum::Eqiation=...ij, ...jk -> ...ik, InputSize=2, OutputSize=1, InputShape={{47, 51}, {51, 83}}               0.523          0.292           1.79     
einsum::Layer_Einsum::Eqiation=ij -> i, InputSize=1, OutputSize=1, InputShape={{30, 40}}                                       0.027          0.026           1.03     
einsum::Layer_Einsum::Eqiation=ij -> i, InputSize=1, OutputSize=1, InputShape={{113, 374}}                                     0.121          0.120           1.01     
einsum::Layer_Einsum::Eqiation=ij, jk -> ik, InputSize=2, OutputSize=1, InputShape={{2, 3}, {3, 2}}                            0.058          0.053           1.08     
einsum::Layer_Einsum::Eqiation=ij, jk -> ik, InputSize=2, OutputSize=1, InputShape={{20, 30}, {30, 20}}                        0.135          0.119           1.14     
einsum::Layer_Einsum::Eqiation=ij, jk -> ik, InputSize=2, OutputSize=1, InputShape={{113, 127}, {127, 113}}                    3.635          2.044           1.78     
einsum::Layer_Einsum::Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 7, 9}, {1, 5, 9, 8}}         0.118          0.125           0.95     
einsum::Layer_Einsum::Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 70, 90}, {1, 5, 90, 80}}     30.297         10.084          3.00     
einsum::Layer_Einsum::Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 73, 91}, {1, 5, 91, 57}}     23.522         8.031           2.93  
```

asmorkalov commented:
x86 without AVX2:
```
Geometric mean (ms)

                                                       Name of Test                                                        4.x-baseline-1 4.x-fastgemm-1 4.x-fastgemm-1
                                                                                                                                                               vs      
                                                                                                                                                         4.x-baseline-1
                                                                                                                                                           (x-factor)  
einsum::Layer_Einsum::Eqiation=...ij -> ...i, InputSize=1, OutputSize=1, InputShape={{30, 40}}                                 0.007          0.007           0.96     
einsum::Layer_Einsum::Eqiation=...ij -> ...i, InputSize=1, OutputSize=1, InputShape={{113, 374}}                               0.043          0.044           1.00     
einsum::Layer_Einsum::Eqiation=...ij, ...jk -> ...ik, InputSize=2, OutputSize=1, InputShape={{40, 50}, {50, 80}}               0.254          0.153           1.66     
einsum::Layer_Einsum::Eqiation=...ij, ...jk -> ...ik, InputSize=2, OutputSize=1, InputShape={{47, 51}, {51, 83}}               0.289          0.163           1.77     
einsum::Layer_Einsum::Eqiation=ij -> i, InputSize=1, OutputSize=1, InputShape={{30, 40}}                                       0.007          0.007           0.97     
einsum::Layer_Einsum::Eqiation=ij -> i, InputSize=1, OutputSize=1, InputShape={{113, 374}}                                     0.042          0.044           0.97     
einsum::Layer_Einsum::Eqiation=ij, jk -> ik, InputSize=2, OutputSize=1, InputShape={{2, 3}, {3, 2}}                            0.007          0.007           1.10     
einsum::Layer_Einsum::Eqiation=ij, jk -> ik, InputSize=2, OutputSize=1, InputShape={{20, 30}, {30, 20}}                        0.038          0.030           1.28     
einsum::Layer_Einsum::Eqiation=ij, jk -> ik, InputSize=2, OutputSize=1, InputShape={{113, 127}, {127, 113}}                    0.760          0.630           1.21     
einsum::Layer_Einsum::Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 7, 9}, {1, 5, 9, 8}}         0.050          0.044           1.14     
einsum::Layer_Einsum::Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 70, 90}, {1, 5, 90, 80}}     4.641          3.224           1.44     
einsum::Layer_Einsum::Eqiation=imkj, injs -> imnks, InputSize=2, OutputSize=1, InputShape={{1, 4, 73, 91}, {1, 5, 91, 57}}     3.735          2.685           1.39     
```

asmorkalov commented:
So fastGemm integration totally makes sense. I propose to  Extract platform detection and run it once in layer constructor. Other things looks good to me.

Abdurrahheem commented:
> So fastGemm integration totally makes sense. I propose to Extract platform detection and run it once in layer constructor. Other things looks good to me.

In my last comment I have fixed tests issues mentioned by @dkurt and fixed platform detection. Should I get results with new platform detection or can we merge the PR without it. In my opinion there will not be too much of a difference in terms of results, so we can merge the PR

asmorkalov commented on modules/dnn/src/layers/einsum_layer.cpp:
```diff
@@ -1389,6 +1294,114 @@ Mat LayerEinsumImpl::pairwiseOperandProcess(
+            reshapedInput2 = input2.reshape(1, 2, shape2);
+        }
+
+        Mat tmp_output;
+        tmp_output.create(M, N, reshapedInput1.type());
```
You can do it in single line:
```
Mat tmp_output(M, N, reshapedInput1.type());
```

asmorkalov commented on modules/dnn/src/layers/einsum_layer.cpp:
```diff
@@ -1389,6 +1294,114 @@ Mat LayerEinsumImpl::pairwiseOperandProcess(
+            int shape2[] = {static_cast<int>(K), static_cast<int>(N)};
+            part2 = part2.reshape(1, sizeof(shape2)/sizeof(shape2[0]), shape2);
+
+            Mat tmp_output;
+            tmp_output.create(M, N, part1.type());
```
You can do it in single line:
```
Mat tmp_output(M, N, part1.type());
```

asmorkalov commented:
@dkurt Do you have other remarks? If no, I propose to merge the PR after constructors fix.

fengyuentau commented on modules/dnn/src/layers/einsum_layer.cpp:
```diff
@@ -6,6 +6,8 @@
 #include "../precomp.hpp"
 #include "layers_common.hpp"
+#include "cpu_kernels/fast_gemm.hpp"
+#include "cpu_kernels/fast_gemm.hpp"
```
two includes?

asmorkalov reviewed (approved):
:+1:

dkurt commented on modules/dnn/src/layers/einsum_layer.cpp:
```diff
@@ -1389,6 +1293,112 @@ Mat LayerEinsumImpl::pairwiseOperandProcess(
     return output;
 };
 
+Mat LayerEinsumImpl::batchwiseMatMul(
```
Are there any changes in this func? I propose to reduce number of changes and keep implementation in the top of the file (adding `LayerEinsumImpl::`)

asmorkalov replied:
I checked difference. It's related to fastGemm only.

dkurt commented on modules/dnn/src/layers/einsum_layer.cpp:
```diff
@@ -1389,6 +1293,112 @@ Mat LayerEinsumImpl::pairwiseOperandProcess(
+        // check if input1 needs reshape, if need reshape
+        if (input1.size[0] != M || input1.size[1] != K)
+        {
+            int shape[] = {static_cast<int>(batches), static_cast<int>(M), static_cast<int>(K)};
```
All the entries with `M`, `K`, `N` use `static_cast`. Can they be initialized as `int`?

Abdurrahheem replied:
changed

dkurt commented on modules/dnn/src/layers/einsum_layer.cpp:
```diff
@@ -1389,6 +1293,112 @@ Mat LayerEinsumImpl::pairwiseOperandProcess(
+    for (size_t i = 0; i < output.size(); i++) {
+        Mat output_slice = output_buffer.row(i);
+        output[i].copyTo(output_slice);
+    }
```
`hconcat`?

Abdurrahheem replied:
Tried using concat. For some reason it fails the inference. Can you suggest the showcase the usage you had in your mind?

asmorkalov replied:
What is error message?

dkurt replied:
Excuse me, `vconcat` for sure:
```cpp
// ...
output.emplace_back(tmp_output.reshape(1, 1));
// ...
Mat output_buffer;
cv::vconcat(output, output_buffer);

int outputDim[] = {static_cast<int>(output.size()), M, N};
output_buffer = output_buffer.reshape(1, 3, &outputDim[0]);
```

dkurt commented:
@Abdurrahheem, branch was pushed to origin by mistake: https://github.com/opencv/opencv/tree/ash/dev_einsum_fast_gemm

Please do locally:
```
git remote set-url --push origin ""
```
To avoid pushing to OpenCV:
```
$ git remote -v
dkurt   https://github.com/dkurt/opencv (fetch)
dkurt   https://github.com/dkurt/opencv (push)
origin  https://github.com/opencv/opencv (fetch)
origin   (push)
```
